Documenting rollback in faststore (#457)
vtex/faststore: docs/wip/how-to-rollback.md

> ## ⚠️ Warning: this will only work when we start using VTEX infrastructure to build the images. Is also possible to build with Netlify[] — for rollback in Netlify, you can use it's tools.

Reviewer: ```suggestion
## ⚠️ Warning: this will only work when we start using VTEX infrastructure to build the images. Is also possible to build with Netlify[] — for rollback in Netlify, you can use its tools.
```

> ## ⚠️ Warning: this will only work when we start using VTEX infrastructure to build the images. Is also possible to build with Netlify[] — for rollback in Netlify, you can use it's tools.

Reviewer 2: ```suggestion
>⚠️ ***Warning:** The rollback process will be available when we start using VTEX infrastructure to build images. For now, we suggest using [Netlify](https://docs.netlify.com/configure-builds/get-started/) to build images and its rollback tools.*
```

# The rollback

> The rollback interaction aims to help you to react faster when having trouble, changing the deploy that is on production thought a safe and tested path. This also allows you to reuse older versions of a storefront, for a recurrent store's campaing for example. 

Reviewer: ```suggestion
The rollback interaction aims to help you to react faster when having trouble, changing the deploy that is on production through a safe and tested path. This also allows you to reuse older versions of a storefront, for a recurrent store's campaign for example. 
```

# 📚 Step by step on how to rollback
This document is the first version of a guide for the rollback process using VTEX Store Framework Jamstack. It also explains the rollback expected behavior.
We'll start showing how to make a safe deploy — one that is ready to be used for a rollback.

> ## 👮‍♀️ How to make a safe deploy (Security Measures)

Reviewer: ```suggestion
## 👮‍♀️ How to make a safe deployment (Security Measures)
```

> # 📚 Step by step on how to rollback
> This document is the first version of a guide for the rollback process using VTEX Store Framework Jamstack. It also explains the rollback expected behavior.
> We'll start showing how to make a safe deploy — one that is ready to be used for a rollback.
> 
> ## 👮‍♀️ How to make a safe deploy (Security Measures)
> - Create a new branch from the current master (this is a backup branch):

Reviewer 2: ```suggestion
# Step by step

## Step 1: Making a safe deploy

In case you need to do a rollback in the future, we always recommend you create a backup branch of the `master` as a security measure. For that, take the following steps:

1.  Create a backup branch from the current master:

```

```
git checkout master
git pull
git checkout -b backup-branch
```

> - Merge your PR, this will trigger a new version to go up.

Reviewer 2: ```suggestion

>ℹ️ *In this example, we named the backup branch as `backup-branch`. However, feel free to name it however you want.*

2. Merge your PR. This will automatically trigger a new version to go up.
```

> ## 🔙 Rolling back
> If you need to rollback after a deploy, you now have a backup branch named `backup-branch`. 
> An example curl is presented below:

Reviewer 2: ```suggestion
## Step 2: Rolling back

If for any reason, you need to rollback after a deploy, you can do it safely as now we have a backup branch. 

To roll back to the `backup-branch`, you must send a POST request to SFJ APIs as in the following:

>⚠️ ***Keep in mind:** Due to internal updates, the URL `http://builders.staging-2a.ingress.vtex.io` may change in the future.*

```

```
curl --location --request POST 'http://builders.staging-2a.ingress.vtex.io' \
--header 'Content-Type: application/json' \
--header 'Authorization: <VTEX IO token>' \
--data-raw '{
    "account": "<account>",
    "action": "rollback",
     "head_commit": {

>         "id": "<commit ID>"

Reviewer 2: ```suggestion
        "id": "<commit-ID>"
```

    },
    "repository": {

>         "name": "<repository Name>",

Reviewer 2: ```suggestion
        "name": "<repository-name>",
```

        "owner": {
            "login": "vtex-sites"
        }
    },
    "target": {

>         "branch": "backup-branch"

Reviewer 2: ```suggestion
        "branch": <backup-branch>
```

    }
}'
```

> - You can get the `VTEX IO token` in the toolbelt, running the command `vtex local token`;
> - The `commit ID` is the hash that identifies the head commit of the branch `backup-branch`;
> - The repository name is the name of your store's GitHub project, for example `storeName.store`.

Reviewer 2: ```suggestion
- `VTEX IO token` -  The necessary authentication code to send requests to VTEX APIs. Run `vtex local token` in the terminal to obtain an exclusive token.
- `account` (string) - The VTEX account name.
- `commit-ID` (string) - The hash that identifies the head commit of the `backup-branch` branch.
- `repository-name` (string)- The name of your store's GitHub project (e.g., `storecomponents.store`).
- `backup-branch` (string) - The backup branch name (e.g., `backup-branch`).
```

> Wait for the `2xx` reponse.

Reviewer 2: ```suggestion
Once you send the POST request, wait for the `2xx` HTTP status code.
```

> # Expected Behavior
> When the rollback is performed with a `2xx` status response, it triggers a new build of your store from the branch `backup-branch`.
> Because of that, the time to replace the current store can take about 10 minutes, depending on the size of your store.
> 
> All the subsequent production builds of your store will be based on that commit from the `backup-branch`.
> Let's suppose you where fixing a bug, than after finishing it you and you want to go back to the previously `main` version of your store. Than, you need to perform the request:

Reviewer 2: ```suggestion
# Modus Operandi

Once the rollback is successfully performed, the `2xx` HTTP status will trigger a new build from the backup branch. 

>⚠️ *Depending on your store size, the time to replace the build in production can take about 10 minutes.*

All the subsequent production builds of your store will be based on that commit from the backup branch.

Therefore, suppose you are fixing a bug. After finishing it, you decide to go back to the previous `main` version of your store. Then, you must perform the following request:

```

```
curl --location --request POST 'http://builders.staging-2a.ingress.vtex.io' \
--header 'Content-Type: application/json' \
--header 'Authorization: <VTEX IO token>' \
--data-raw '{
    "account": "<account>",
    "action": "rollback",
     "head_commit": {
        "id": "<commit ID>"
    },
    "repository": {
        "name": "<repository Name>",
        "owner": {
            "login": "vtex-sites"
        }
    },
    "target": {
        "branch": "main"
    }
}'
```
If your default branch is called `master` instead of `main`, this will perform a production build with the code in the `main` branch.

> It will also allow new builds from the `main` branch everytime a merge happens.

Reviewer: ```suggestion
It will also allow new builds from the `main` branch every time a merge happens.
```

> If your default branch is called `master` instead of `main`, this will perform a production build with the code in the `main` branch.
> It will also allow new builds from the `main` branch everytime a merge happens.

Reviewer 2: ```suggestion
>⚠️ *Notice that if your default branch is called `master` instead of `main`, the previous request will run a production build with the code in the main branch. It will also allow new builds from the main branch every time a merge happens.*
```

> ## Technical details

Reviewer 2: ```suggestion
```

> You don't actually need to use the branch `backup-branch`, you can name it however you want.

Reviewer: ```suggestion
You don't need to use the branch `backup-branch`, you can name it however you want.
```

> There is the possibility of the url `http://builders.staging-2a.ingress.vtex.io` changing due to infrastructure updates.

Reviewer: ```suggestion
There is the possibility of the URL `http://builders.staging-2a.ingress.vtex.io` changing due to infrastructure updates.
```

> You don't actually need to use the branch `backup-branch`, you can name it however you want.
> 
> There is the possibility of the url `http://builders.staging-2a.ingress.vtex.io` changing due to infrastructure updates.
> 
> For better clarity, here is the format of the body:
> ```
> {
>     "account": string,
>     "action": "rollback",
>     "head_commit": {
>         "id": string
>     },
>     "repository": {
>         "name": string,
>         "owner": {
>             "login": "vtex-sites"
>         }
>     },
>     "target": {
>         "branch": string
>     }
> }
> ```
> 

Reviewer 2: ```suggestion
```

# FAQ

> ### How much time it takes?

Reviewer: ```suggestion
### How much time does it take?
```

We are working to have a fast rollback, the time may not be longer than 10 minutes.

### How to take security measures? 
Always check the deploy preview before putting a deploy on production. The rollback interaction will invite you to do so. 

### How much time it takes to invalidate caching

> We are investigatig it by now. 

Reviewer: ```suggestion
We are investigating it by now. 
```

- Caching? How much time it takes to invalidate caching

# Expected Behavior
When the rollback is performed with a `2xx` status response, it will trigger a new build of your store from the branch `backup-branch`.
Because of that, the time to replace the current store can take about 10 minutes, depending on the size of your store.

All the subsequent production builds of your store will be based on that commit from the `backup-branch`.

> Once the bug is fixed, to go back to the `main` version of your store, you need to perform the request:

Reviewer 3: Isn't making this request after the bug is fixed in the main branch redundant since the merge will update the target branch, as mentioned below:

```
It will also allow new builds from the `main` branch everytime a merge happens.
```

And some questions:
- what kind of authorization is done when calling this API? specifically, what happens if I provide an io token from account X to update the target branch/commit of account Y?
- If the target branch is `backup-branch` and someone merges another branch into it or pushes a commit, will that trigger a build? (I think the current behavior will change depending on whether the incoming commit belongs to a PR) If so, will the master route (i.e marinbrasil.vtex.app) point to the new deploy preview or would the user need to call the API again after pushing to `backup-branch`

Author: I'm just checking about when the bug is fixed in the default branch, but I'm confident you're right

I'll add the details of your questions to the documentation. But I'll leave it also here for reference
- When the authorization fails, the response status is `4xx`. The rollback will take place when the response status is `201`. The token comes from `vtex local token` and the autorization is the one with VTEX ID, checking whether the token is capable of accessing the `workspace-read-write` resource. I've found no part of the code that protects from cross-account rollback. I'm adding it to our backlog
- There is no logic for that in place. There can be as many pushes to the non-default branch where the rollback happened, and it will not change

I'm just checking some details, I'll rework this part of the documentation

```
curl --location --request POST 'http://builders.staging-2a.ingress.vtex.io' \
--header 'Content-Type: application/json' \
--header 'Authorization: <VTEX IO token>' \
--data-raw '{
    "account": "<account>",
    "action": "rollback",
     "head_commit": {
        "id": "<commit ID>"
    },
    "repository": {
        "name": "<repository Name>",
        "owner": {
            "login": "vtex-sites"
        }
    },
    "target": {
        "branch": "main"
    }
}'
```
Perhaps your default branch is called `master` instead of `main`.
This will perform a production build with the code in the `main` branch.
It will also allow new builds from the `main` branch everytime a merge happens.
## Technical details
You don't actually need to use the branch `backup-branch`, you can name it however you want.

> There is the possibility of the url `http://builders.staging-2a.ingress.vtex.io` changing due to infrastructure updates.

Reviewer 2: @dhasuda Coloquei na linha 26 das sugestões que eu fiz. Em que sentido você não sabe como falar disso? Na verdade, eu não sei se há necessidade de incluir isso, porque o ideal é a doc sempre estar atualizada

> There is the possibility of the url `http://builders.staging-2a.ingress.vtex.io` changing due to infrastructure updates.

Author: Hoje fizemos um update e estabilizamos a URL. Antes essa URL iria mudar quase sempre que tivéssemos algum problema na nossa infraestrutura, agora conseguimos fazer algo um pouco mais perene
